index.js

```diff
@@ -18,311 +16,98 @@ var ONE_DAY = 24 * 60 * 60 * 1000;
  */
 
-module.exports = function(opts, app){
+module.exports = function(opts, app) {
   // session(app[, opts])
   if (opts && typeof opts.use === 'function') {
-    var tmp = app;
+    const tmp = app;
     app = opts;
     opts = tmp;
   }
-
-  opts = opts || {};
-
-  // key
-  opts.key = opts.key || 'koa:sess';
-
-  // back-compat maxage
-  if (!('maxAge' in opts)) opts.maxAge = opts.maxage;
-
-  // defaults
-  if (null == opts.overwrite) opts.overwrite = true;
-  if (null == opts.httpOnly) opts.httpOnly = true;
-  if (null == opts.signed) opts.signed = true;
-
-  debug('session options %j', opts);
-
+  // app required
   if (!app || typeof app.use !== 'function') {
     throw new TypeError('app instance required: `session(opts, app)`');
   }
 
-  // setup encoding/decoding
-  if (typeof opts.encode !== 'function') {
-    opts.encode = encode
-  }
-  if (typeof opts.decode !== 'function') {
-    opts.decode = decode
-  }
-
-  // to pass to Session()
-  app.context.sessionKey = opts.key;
-
-  app.context.__defineGetter__('session', function(){
-    var sess = this._sess;
-    // already retrieved
-    if (sess) return sess;
-
-    // unset
-    if (false === sess) return null;
-
-    var json = this.cookies.get(opts.key, opts);
-
-    if (json) {
-      debug('parse %s', json);
-      try {
-        // make sure sessionOptions exists
-        initSessionOptions(this, opts);
-        var obj = opts.decode(json);
-        if (typeof opts.valid === 'function' && !opts.valid(this, obj)) {
-          // valid session value fail, ignore this session
-          sess = new Session(this);
-          json = obj;
-          debug('invalid %j', obj);
-        } else {
-          sess = new Session(this, obj);
-          // make prev a different object from sess
-          json = opts.decode(json);
-        }
-      } catch (err) {
-        // backwards compatibility:
-        // create a new session if parsing fails.
-        // new Buffer(string, 'base64') does not seem to crash
-        // when `string` is not base64-encoded.
-        // but `JSON.parse(string)` will crash.
-        debug('decode %j error: %s', json, err);
-        if (!(err instanceof SyntaxError)) throw err;
-        sess = new Session(this);
-        json = null;
-      }
-    } else {
-      debug('new session');
-      sess = new Session(this);
-    }
-
-    this._sess = sess;
-    this._prevjson = json;
-    return sess;
-  });
-
-  app.context.__defineSetter__('session', function(val){
-    if (null == val) return this._sess = false;
-    if ('object' == typeof val) return this._sess = new Session(this, val);
-    throw new Error('this.session can only be set as null or an object.');
-  });
+  opts = formatOpts(opts);
+  extendContext(app.context, opts);
 
-  return function* session(next){
-    // make sessionOptions independent in each request
-    initSessionOptions(this, opts);
+  return function* session(next) {
+    if (this.sess.store) yield this.sess.initFromExternal();
```

Comment:
cookie store is a little different from other external stores.

1. it is synchronize so it can be lazy loading,
2. it use the same cookie field to store the value.

If we treat cookie session as a store, we still need to write some tricky code.